spacy/tokens/doc.pyx

```diff
@@ -1743,6 +1743,127 @@ cdef class Doc:
+            in ascending order. For example, if *ps_lengths==[1, 2]*, *ps_search=="aC" and *cs==False*, the searched strings 
+            hashed for "spaCy" would be "a" and "ac".
+        ss_search_chars: a byte array containing, in numerical order, UTF-8 characters to search for within each token, 
+            starting at the end.
+        ss_width_offsets: an array of single-byte values [1-char-start, 2-char-start, 3-char-start, 4-char-start, 4-char-end]
+            specifying the offsets within *ss_search_chars* that contain UTF-8 characters with the specified widths.
+        ss_lengths: an array of single-byte values specifying the lengths of search results (from the end) to be hashed
+             in ascending order. For example, if *ss_lengths==[1, 2]*, *ss_search=="aC" and *cs==False*, the searched strings 
+             hashed for "spaCy" would be "c" and "ca".
+    
+        Many of the buffers passed into and used by this method contain single-byte numerical values. This takes advantage of 
+        the fact that we are hashing short affixes and searching for small groups of characters. The calling code is responsible
+        for ensuring that lengths being passed in cannot exceed 63 and hence that resulting values with maximally four-byte 
+        character widths can never exceed 255.
+
+        Note that this method performs no data validation itself as it expects the calling code will already have done so, and
+        that the behaviour of the code may be erratic if the supplied parameters do not conform to expectations.
+        """
+        
+        # Work out lengths
+        cdef int p_lengths_l = strlen(<char*> p_lengths)
+        cdef int s_lengths_l = strlen(<char*> s_lengths)
+        cdef int ps_lengths_l = strlen(<char*> ps_lengths)
+        cdef int ss_lengths_l = strlen(<char*> ss_lengths)
+        cdef int hashes_per_tok = p_lengths_l + s_lengths_l + ps_lengths_l + ss_lengths_l
+        cdef int p_max_l = p_lengths[p_lengths_l - 1] if p_lengths_l > 0 else 0
+        cdef int s_max_l = s_lengths[s_lengths_l - 1] if s_lengths_l > 0 else 0
+        cdef int ps_max_l = ps_lengths[ps_lengths_l - 1] if ps_lengths_l > 0 else 0
+        cdef int ss_max_l = ss_lengths[ss_lengths_l - 1] if ss_lengths_l > 0 else 0
+        
+        # Define / allocate buffers
+        cdef Pool mem = Pool()
+        cdef unsigned char* pref_l_buf = <unsigned char*> mem.alloc(p_max_l, 1)
+        cdef unsigned char* suff_l_buf = <unsigned char*> mem.alloc(s_max_l, 1)
+        cdef unsigned char* ps_res_buf = <unsigned char*> mem.alloc(ps_max_l, 4)
+        cdef unsigned char* ps_l_buf = <unsigned char*> mem.alloc(ps_max_l, 1)
+        cdef unsigned char* ss_res_buf = <unsigned char*> mem.alloc(ss_max_l, 4)
+        cdef unsigned char* ss_l_buf = <unsigned char*> mem.alloc(ss_max_l, 1)
+        cdef int doc_l = self.length, total_hashes = doc_l * hashes_per_tok 
+        cdef np.uint64_t* hashes_ptr = <np.uint64_t*> mem.alloc(
```

Comment:
Any reason this buffer can't be the `numpy` array itself? We could call `numpy.zeros` with `total_hashes` and use its `data` attribute to perform the operations in-place, and we save a copy in the process.

Reply:
Good point! I've changed it accordingly (using `numpy.empty` rather than `numpy.zeros` as all the positions in the buffer are guaranteed to be written to).